lib/block-supports/border.php

```diff
@@ -125,6 +125,25 @@ function gutenberg_skip_border_serialization( $block_type ) {
 		$border_support['__experimentalSkipSerialization'];
 }
 
+/**
+ * Checks whether the current block type supports the border feature requested.
+ *
+ * @param WP_Block_Type $block_type Block type to check for support.
+ * @param string        $feature    Name of the feature to check support for.
+ * @param mixed         $default    Fallback value for feature support, defaults to false.
+ *
+ * @return boolean                  Whether or not the feature is supported.
+ */
+function gutenberg_has_border_feature_support( $block_type, $feature, $default = false ) {
+	$block_support = false;
+	if ( property_exists( $block_type, 'supports' ) ) {
+		$block_support = _wp_array_get( $block_type->supports, array( '__experimentalBorder' ), $default );
```

Comment:
> This is: gutenberg_block_has_support is true both for __experimentalBorder: true and __experimentalBorder: { somePropIsEnabled }, which is problematic in this case, so we can't rely on it when checking for a specific prop.

If I'm understanding your comment correctly, yes, we cannot rely on `gutenberg_block_has_support`. It only checks the provided path specifically. 

> It seems like what we want here is to check whether __experimentalBorder is boolean and true, otherwise, this is going to give us false positives, correct?

Maybe I've gone in circles a few too many times and gotten confused. This is what the code above does.
It retrieves the `__experimentalBorder` support if available. It checks that for boolean `true` then failing that, checks for the specific feature using `gutenberg_block_has_support`.

In case it helps clarify the discussion, we have two scenarios that equate to the feature support being opted into.

- All border features are opted into via `__experimentalBorder` being set to `true`
- The individual feature is opted into via a nested flag under `__experimentalBorder`

Currently, `gutenberg_block_has_support` checks a specific path, that is, it only retrieves the value for that path and checks whether it is `true` or is an array.

So if we check for a path such as `array( '__experimentalBorder', 'color' )` it will only be checking if `__experimentalBorder.color` is `true` or an array. If `__experimentalBorder` itself was set to boolean `true`, `gutenberg_block_has_support` returns the `$default` value. In this case giving us a false negative.

This new function determines the border feature support to be available if `__experimentalBorder` is set to boolean `true` otherwise returns the feature specific results via `gutenberg_block_has_support`.

`gutenberg_block_has_support` could be updated to return `true` if any node along the path was set to boolean `true`. I thought this change too wide reaching if other block supports did not wish to allow such blanket opt in to individual features.

Reply:
Oh, it tripped me up, sorry! This is working as expected. However, both of us seem to have struggled with it :sweat_smile: which sounds like a smell to stop for a minute and think about our future selves. What do you think of refactoring this function along the lines of:

```php
$block_support = false;
if (
  property_exists( $block_type, 'supports' ) &&
  ( true === _wp_array_get( $block_type->supports, array( '__experimentalBorder' ), $default ) )
) {
  $block_support = true;
}

return $block_support || gutenberg_block_has_support( $block_type, array( '__experimentalBorder', $feature ), $default );
```

That structure gives me more confidence about our future selves being able to understand what this code does. Thoughts?

Reply:
Always happy to make the code's intent clearer and save anyone effort in the future.

I've updated this check along the suggested lines in https://github.com/WordPress/gutenberg/pull/31020/commits/b5b08fa2f940585ab0f10d1bba4e179ff45eab41. I took it a step further to return immediately if blanket opt-in has been made. You'd likely need to grok the conditional anyway to see how `$block_support` was set. Hopefully that isn't taking it a step too far.

In addition, I added inline comments and improved the functions doc block comments.

Reply:
That's great, thanks!